logic/resolvable/Retrievable.java

```diff
@@ -35,4 +76,196 @@ public Retrievable asRetrievable() {
     public boolean isRetrievable() {
         return true;
     }
-}
\ No newline at end of file
+
+    public static class Extractor {
+        private final Conjunction conjunction;
+        private final Set<Concludable<?>> concludables;
+
+        private final Map<Variable, Variable> variablesMap = new HashMap<>(); // Map from variables in original conjunction to new subConjunction
+        private final Set<Constraint> visitedConstraints = new HashSet<>();
+        private final Set<Set<Variable>> connectedSubgraphs = new HashSet<>();
+        private Set<Variable> connectedSubgraph;
+
+
+        public Extractor(Conjunction conjunction, Set<Concludable<?>> concludables) {
+            this.conjunction = conjunction;
+            this.concludables = concludables;
+        }
+
+        public static Extractor from(Conjunction conjunction, Set<Concludable<?>> concludables) {
+            return new Extractor(conjunction, concludables);
+        }
+
+        public Set<Retrievable> extract() {
+            concludables.forEach(concludable -> visitedConstraints.addAll(concludable.coreConstraints()));
+            conjunction.variables().stream().filter(var -> var.identifier().reference().isName()).forEach(var -> {
+                if (variablesMap.get(var) == null) {
+                    connectedSubgraph = new HashSet<>();
+                    if (var.isThing()) addThingVariable(var.asThing());
+                    else if (var.isType()) addTypeVariable(var.asType());
+                    else throw GraknException.of(ILLEGAL_STATE);
+                    connectedSubgraphs.add(connectedSubgraph);
+                }
+            });
+            return connectedSubgraphs.stream()
+                    .filter(subGraph -> !(subGraph.size() == 1 && subGraph.iterator().next().constraints().size() == 0))
+                    .map(subGraph -> new Retrievable(new Conjunction(subGraph, set()))).collect(Collectors.toSet());
+        }
+
+        private void addToConnected(Variable copy) {
+            connectedSubgraph.add(copy);
+        }
+
+        private TypeVariable addTypeVariable(TypeVariable variable) {
+            return addVariable(variable).asType();
+        }
+
+        private ThingVariable addThingVariable(ThingVariable variable) {
```

Comment:
hmmm okay this led me down a train of thought:
1. each retrievable should be its own subgraph
2. conceptually, if it were possible for `$x` to be present in two different subgraphs, we would have a bug because the `variablesMap` is not re-instantiated per subgraph. Even if the subgraphs can never share a variable, its a bit unclear that the `variablesMap` holds variables remapped from different subgraphs?

I think its a tad confusing/smelly because you have a stateful "subgraph" that you add to during a single recursion, and then re-instantiate.

Possibly a cleaner pattern is:
you have a `SubgraphRegister` that you instantiate. You pick your starting variable, then do `subgraphRegister.registerVariable(variable)`. This will do all the registering of variables _within_ the `SubgraphRegister` object, which solves the problem/conceptual oddity in point 2 above. Now, each subgraph has its own `VariablesMap` that is only uses internally.
After the recursion ends you add all the variables and constraints registered in this subgraph into a global `visitedVariables`/`visitedConstraints`.

Does that make sense, not sure if unclear in explaining lol

Reply:
That does make sense, however to avoid infinite recursion we need to update the visited variables and constraints as we recurse, but I think these can be held locally to the subgraph, and then added to the global visited maps once that subgraph is complete.